### ISSUE TITLE
REST API failing with sanitation of JSON response in Arrays

### ISSUE THREAD

**Issue by JPchico** [aiidateam/aiida-core#5559]
### Describe the bug

An AiiDA array might contain a NaN as one cannot store None values in np.arrays. This causes that when the array is returned via the REST-API the JSON will not have the correct format as NaN are not acceptable JSON entities, these should be changed to `null`

### Steps to reproduce

Store an `orm.ArrayData` with a `NaN`, try to download it using the REST-API.

### Expected behavior

When generating the dictionary that will be used for the JSON download the `NaN` should be sanitized to `null`

### Your environment

- Operating system [e.g. Linux]: 18.04
- Python version [e.g. 3.7.1]: 3.8.11
- aiida-core version [e.g. 1.2.1]: AiiDA v2.0.1.post0

### Additional context

When I first did a simple implementation of this #5166 I had to set `ignore_nan=True` in the `json.dumps` to ensure that the `Nan` were transformed to `null` and avoid this problem. In #5425 this was removed, I'm unsure if it is because it is assumed that there should be no `NaN` in the Arrays. However, in case they exist one should account for this in the serving of the data. 

If `NaN` should not be allowed in the `orm.ArrayData` I think that there should be a check that stops the user from storing these values.

**Comment by JPchico** [comment]
Looking at this away, it seems to be that the issue comes from the fact that the `simplejson` dependency was removed and the base `json` package is used instead. This means that the `ignore_nan=True` option is not available anymore.

I think that the best solution for now would be to have a custom function that sanitizes the array when it is going to be displayed as a json. Basically it would take any `NaN` -> `null`. It should then be possible if one wants to then very simply add a function that would read any `null` in an `ArrayData` as a `NaN`.

The situation is more complex if one wants to deal with `np.inf` or something similar. 

A quick and dirty way could be the following:

```python
shape_array = np.shape(array_data)
_clean_flat_list = [entry if not np.isnan(entry) else None for entry in shape_array.flatten().tolist()]
clean_list = np.reshape(np.asarray(_clean_flat_list), shape).tolist()
```

If you agree with this I can add this in

https://github.com/aiidateam/aiida-core/blob/3e4c8837aa6a4472482465177a13ed00c99d6dd0/aiida/orm/nodes/data/array/array.py#L204-L206

where to each `val` one would apply the function that I describe. The tricky thing is to decode from json to `ArrayData` I do not know which method would do that. 

What do you think @sphuber and @chrisjsewell do you agree?

**Comment by sphuber** [comment]
I think that since `NaN` exists in Python and `numpy` handles this just fine, and `ArrayData` simply stores `numpy` arrays, the current behavior of allowing it makes sense, and so I would keep it. The following works and I would keep it working:
```python
In [17]: a = ArrayData()

In [18]: a.set_array('array', numpy.array([float('NaN'), float('inf'), float('-inf')]))

In [19]: a.get_array('array')
Out[19]: array([ nan,  inf, -inf])

In [20]: a.store()
Out[20]: <ArrayData: uuid: adf6582c-b64d-47a9-a91e-3a20aede8c79 (pk: 57997)>
```
So I agree with your suggested approach that we should simply update `ArrayData._prepare_json` to deal with `NaN` and `inf` values and replace them with `null`. Setting `NaN` to `null` makes sense but apparently the ECMA standard also specified this for `inf`. See [this pdf](https://www.ecma-international.org/wp-content/uploads/ECMA-262_12th_edition_june_2021.pdf) on page 739 which states:

> inite numbers are stringified as if by calling ToString(number). NaN and Infinity regardless of sign are represented as the String "null".

This may be confusing to consumers of the REST API as they won't be able to distinguish between `None`, `NaN` and `inf`, but at least it is correct JSON. We should probably just document it as it is the best we can do as long as we allow `NaN` and `inf` in `ArrayData` nodes.

**Comment by JPchico** [comment]
@sphuber thanks for the answer. I think if that is what the standard requires that is what we should do. It is of course not ideal that we cannot identify between `np.nan` and `np.inf` when consuming the JSON. However, I think that in most cases if you consume an array that has `np.nan` and/or `np.inf` what you are interested is in the finite entries, and in knowing which entries are non-finite. In this case rendering them as `null` should not cause a big problem.

I'll be making a small PR to address this issue, with the check being expanded to include `np.inf` and `-np.inf`

**Comment by JPchico** [comment]
I have made a small PR on this, basically it is the same function that I described before but I also included the check `np.isinf` so those are also transformed to `None`

**Comment by sphuber** [comment]
Fixed by #5613
